### PR TITLE
Document evolution from 1- to 0-based indexing

### PR DESCRIPTION
**Pull request by bradcray**
This adds a section to the language evolution document about our decision to switch from 1- to 0-based indexing.  It includes a description of the change, history and rationale, tips for updating code, and offers to help.

### CODE REVIEW

**Comment by bradcray** [comment]
@ben-albrecht: This is ready for your read-through and feedback.

File: doc/rst/language/evolution.rst [hunk 1]
```diff
@@ -3,18 +3,225 @@
...
+Like any language, Chapel has changed over time. This page is designed
+to describe significant language changes that have a high likelihood
+of breaking existing user codes or code samples from presentations or
+papers that predated the changes.
 
 Note that the compiler flag ``--warn-unstable`` is available and can be
 useful when migrating programs to the current version of the language.
-The purpose of this flag is to identify  portions of a program that use a
-language feature that has changed meaning.
+The purpose of this flag is to identify portions of a program that use a
```
Comment (ben-albrecht) [inline doc/rst/language/evolution.rst:13]: Has this flag's meaning changed overtime? I have always thought of it as the latter description (considered unstable and can change in the future).

File: doc/rst/language/evolution.rst [hunk 2]
```diff
@@ -3,18 +3,225 @@
...
+
+0- vs. 1-based Indexing
+***********************
+
+Version 1.22 makes a major change to Chapel with respect to indexing
+for cases that involve implicit indices.  Historically, Chapel has
+used 1-based indexing for such cases, where it now uses 0-based
+indexing.
+
+The major types that are affected by this change are tuples, strings,
```
Comment (ben-albrecht) [inline doc/rst/language/evolution.rst:29]: Does `list` deserve a mention here?

File: doc/rst/language/evolution.rst [hunk 3]
```diff
@@ -3,18 +3,225 @@
...
+
+      var t: 2*int = ...;
+
+      var (x, y) = t;
+
+      for i in t.indices do
+        writeln("t(", i, ") = ", t(i));
+
+  If you have a pattern that you're trying to write in an
+  index-neutral style, but can't, don't hesitate to ask for tips.
```
Comment (ben-albrecht) [inline doc/rst/language/evolution.rst:174]: Should we link https://chapel-lang.org/users.html from here?

File: doc/rst/language/evolution.rst [hunk 4]
```diff
@@ -3,18 +3,225 @@
...
+
+  - Look for any calls on lists that use explicit offsets, as these
+    will likely need updates.  For example ``mylist.pop(1);`` will
+    need to become ``mylist.pop(0);``
+
+Need Help?
+**********
+
+If you are able to share your code with us and would like help
+updating it to Chapel 1.22, please don't hesitate to ask for help.
```
Comment (ben-albrecht) [inline doc/rst/language/evolution.rst:217]: How do we recommend people reaching out? Mailing list?

File: doc/rst/language/evolution.rst [hunk 5]
```diff
@@ -3,18 +3,225 @@
...
+      var (x, y) = t;
+
+      for i in t.indices do
+        writeln("t(", i, ") = ", t(i));
+
+  If you have a pattern that you're trying to write in an
+  index-neutral style, but can't, don't hesitate to ask for tips.
+        
+
+* Some common pitfalls to check for in your code include:
```
Comment (ben-albrecht) [inline doc/rst/language/evolution.rst:177]: Some others that may or may not be worth mentioning:

- Anything that is generating or modifying indices, such as `RandomStream.getNext()` or `ceil()` / `floor()`

- Some common strings to look for are:
  - `[1]`
  - `[2]`
  - `(1)`
  - `(2)`
  - `1..`

File: doc/rst/language/evolution.rst (hunk 1 above)
Comment (bradcray) [reply doc/rst/language/evolution.rst:13]: I would say it has evolved over time.  Historically, we've used it primarily to warn people about things that had already changed, such as `new C()` no longer meaning `new unmanaged C()` (and then no longer meaning `new borrowed C()`).  In this release we added a lot more things in the "be forewarned / may be likely to change" category to it, which is why I updated the wording here.

File: doc/rst/language/evolution.rst (hunk 2 above)
Comment (bradcray) [reply doc/rst/language/evolution.rst:29]: Yeah, I think so.  I always seem to drop them on the floor mentally for some reason.

File: doc/rst/language/evolution.rst (hunk 3 above)
Comment (bradcray) [reply doc/rst/language/evolution.rst:174]: That's a good idea.

File: doc/rst/language/evolution.rst (hunk 4 above)
Comment (bradcray) [reply doc/rst/language/evolution.rst:217]: Let's point to the chapel users page here too if that sounds OK.

Comment (ben-albrecht) [reply doc/rst/language/evolution.rst:217]: Sounds good

File: doc/rst/language/evolution.rst (hunk 5 above)
Comment (bradcray) [reply doc/rst/language/evolution.rst:177]: I like the sentiment of thinking about things that compute indices, but am not sure that getNext, ceil, or floor are necessarily great indicators of such cases (vs. doing some other sort of computation including casting).  So what if we were to say something "Think about whether there are places in your code that compute index values numerically." as a final bullet?

(I like the "common strings" cases, though, and should've thought to put something like that).

Comment (ben-albrecht) [reply doc/rst/language/evolution.rst:177]: > Think about whether there are places in your code that compute index values numerically." as a final bullet?

That works.

**Pull request merged**